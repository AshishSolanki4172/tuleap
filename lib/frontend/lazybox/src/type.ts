/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { HTMLTemplateResult, TemplateResult } from "lit/html.js";
import type { GroupCollection, LazyboxItem } from "./items/GroupCollection";
import type { HTMLTemplateStringProcessor } from "./index";

export interface Lazybox {
    setDropdownContent: (groups: GroupCollection) => void;
    resetSelection: () => void;
    setSelection: (selection: ReadonlyArray<LazyboxItem>) => void;
    destroy: () => void;
}

export type LazyboxSearchFieldCallback = (query: string) => void;

export type LazyboxSelectionCallback = (selected_value: unknown | null) => void;

export type LazyboxTemplatingCallback = (
    html: typeof HTMLTemplateStringProcessor,
    item: LazyboxItem
) => HTMLTemplateResult;

export type LazyboxNewItemCallback = () => void;

export type LazyBoxWithNewItemButton = {
    readonly new_item_button_label: string;
    readonly new_item_callback: LazyboxNewItemCallback;
};

type LazyBoxWithoutNewItemButton = {
    readonly new_item_callback?: undefined;
};

export type LazyboxOptions = (LazyBoxWithoutNewItemButton | LazyBoxWithNewItemButton) & {
    readonly placeholder: string;
    readonly search_input_placeholder: string;
    readonly is_multiple: boolean;
    readonly templating_callback: LazyboxTemplatingCallback;
    readonly selection_callback: LazyboxSelectionCallback;
    readonly search_field_callback: LazyboxSearchFieldCallback;
};

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: unknown;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
}

export interface LazyboxComponent {
    wrapper_element: HTMLElement;
    lazybox_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LazyboxSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface LazyboxSelectionStateMultiple {
    selected_items: Map<string, RenderedItem>;
    selected_values_elements: Map<string, Element>;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}

export interface ManageSelection {
    processSelection(item: Element): void;
    hasSelection(): boolean;
    updateSelectionAfterDropdownContentChange(): void;
    clearSelection(): void;
    setSelection(selection: ReadonlyArray<RenderedItem>): void;
}
