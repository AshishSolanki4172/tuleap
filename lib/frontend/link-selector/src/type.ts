/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { TemplateResult } from "lit/html.js";
import type { LinkSelectorSearchFieldCallback } from "./events/SearchFieldEventCallbackHandler";

export interface LinkSelector {
    destroy: () => void;
}

export interface LinkSelectorOptions {
    readonly placeholder?: string;
    search_field_callback: LinkSelectorSearchFieldCallback;
}

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: string;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
    target_option: HTMLOptionElement;
}

export interface LinkSelectorComponent {
    wrapper_element: HTMLElement;
    link_selector_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LinkSelectorSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}
