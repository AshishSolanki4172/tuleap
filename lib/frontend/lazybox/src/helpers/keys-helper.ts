/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

export function isEscapeKey(event: KeyboardEvent): boolean {
    return event.key === "Escape" || event.key === "Esc";
}

export function isEnterKey(event: KeyboardEvent): boolean {
    return event.key === "Enter";
}

export function isArrowDown(event: KeyboardEvent): boolean {
    return event.key === "ArrowDown";
}

export function isArrowUp(event: KeyboardEvent): boolean {
    return event.key === "ArrowUp";
}

export function isTabKey(event: KeyboardEvent): boolean {
    return event.key === "Tab";
}

export function isBackspaceKey(event: KeyboardEvent): boolean {
    return event.key === "Backspace";
}
