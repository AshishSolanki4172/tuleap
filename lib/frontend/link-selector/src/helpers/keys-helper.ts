/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

export function isEscapeKey(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "Escape" || keyboard_event.key === "Esc";
}

export function isEnterKey(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "Enter";
}

export function isArrowDown(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "ArrowDown";
}

export function isArrowUp(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "ArrowUp";
}

export function isTabKey(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "Tab";
}

export function isShiftKey(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "Shift";
}

function getKeyboardEvent(event: Event): KeyboardEvent {
    if (!(event instanceof KeyboardEvent)) {
        throw new Error("Event is not a keyboard event");
    }

    return event;
}
