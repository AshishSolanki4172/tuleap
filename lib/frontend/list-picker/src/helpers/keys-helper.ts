/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

export function isEscapeKey(event: KeyboardEvent): boolean {
    return event.key === "Escape" || event.key === "Esc" || event.keyCode === 27;
}

export function isBackspaceKey(event: KeyboardEvent): boolean {
    return event.key === "Backspace" || event.keyCode === 8;
}

export function isEnterKey(event: KeyboardEvent): boolean {
    return event.key === "Enter" || event.keyCode === 13;
}

export function isArrowDown(event: KeyboardEvent): boolean {
    return event.key === "ArrowDown" || event.keyCode === 40;
}

export function isArrowUp(event: KeyboardEvent): boolean {
    return event.key === "ArrowUp" || event.keyCode === 38;
}

export function isTabKey(event: KeyboardEvent): boolean {
    return event.key === "Tab" || event.keyCode === 9;
}

export function isShiftKey(event: KeyboardEvent): boolean {
    return event.key === "Shift" || event.keyCode === 16;
}
