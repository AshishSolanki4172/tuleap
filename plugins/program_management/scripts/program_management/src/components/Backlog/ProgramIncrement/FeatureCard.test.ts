/**
 * Copyright (c) Enalean, 2021 - Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { ShallowMountOptions } from "@vue/test-utils";
import { shallowMount } from "@vue/test-utils";
import FeatureCard from "./FeatureCard.vue";
import { createProgramManagementLocalVue } from "../../../helpers/local-vue-for-test";
import type { Feature } from "../../../helpers/ProgramIncrement/Feature/feature-retriever";
import { createStoreMock } from "@tuleap/core/scripts/vue-components/store-wrapper-jest";
import type { ProgramIncrement } from "../../../helpers/ProgramIncrement/program-increment-retriever";

describe("FeatureCard", () => {
    let component_options: ShallowMountOptions<FeatureCard>;

    it("Displays a draggable card with accessibility pattern", async () => {
        component_options = {
            propsData: {
                element: {
                    artifact_id: 100,
                    artifact_title: "My artifact",
                    tracker: {
                        label: "bug",
                        color_name: "lake_placid_blue",
                    },
                    background_color: "peggy_pink_text",
                    has_user_story_planned: false,
                } as Feature,
                program_increment: {
                    user_can_plan: true,
                } as ProgramIncrement,
            },
            localVue: await createProgramManagementLocalVue(),
            mocks: {
                $store: createStoreMock({
                    state: {
                        configuration: { accessibility: true, can_create_program_increment: true },
                    },
                }),
            },
        };

        const wrapper = shallowMount(FeatureCard, component_options);
        expect(wrapper.element).toMatchSnapshot();
    });

    it("Displays a not draggable card without accessibility pattern", async () => {
        component_options = {
            propsData: {
                element: {
                    artifact_id: 100,
                    artifact_title: "My artifact",
                    tracker: {
                        label: "bug",
                        color_name: "lake_placid_blue",
                    },
                    background_color: "",
                    has_user_story_planned: false,
                } as Feature,
                program_increment: {
                    user_can_plan: true,
                } as ProgramIncrement,
            },
            localVue: await createProgramManagementLocalVue(),
            mocks: {
                $store: createStoreMock({
                    state: {
                        configuration: {
                            accessibility: false,
                            can_create_program_increment: false,
                        },
                    },
                }),
            },
        };

        const wrapper = shallowMount(FeatureCard, component_options);
        expect(wrapper.element).toMatchSnapshot();
    });

    it("Displays a not draggable card when feature has planned user stories", async () => {
        component_options = {
            propsData: {
                element: {
                    artifact_id: 100,
                    artifact_title: "My artifact",
                    tracker: {
                        label: "bug",
                        color_name: "lake_placid_blue",
                    },
                    background_color: "",
                    has_user_story_planned: true,
                } as Feature,
                program_increment: {
                    user_can_plan: true,
                } as ProgramIncrement,
            },
            localVue: await createProgramManagementLocalVue(),
            mocks: {
                $store: createStoreMock({
                    state: {
                        configuration: {
                            accessibility: false,
                            can_create_program_increment: true,
                        },
                    },
                }),
            },
        };

        const wrapper = shallowMount(FeatureCard, component_options);
        expect(wrapper.element).toMatchSnapshot();
    });

    it("Displays a not draggable card when user can not plan/unplan features", async () => {
        component_options = {
            propsData: {
                element: {
                    artifact_id: 100,
                    artifact_title: "My artifact",
                    tracker: {
                        label: "bug",
                        color_name: "lake_placid_blue",
                    },
                    background_color: "",
                    has_user_story_planned: true,
                } as Feature,
                program_increment: {
                    user_can_plan: false,
                } as ProgramIncrement,
            },
            localVue: await createProgramManagementLocalVue(),
            mocks: {
                $store: createStoreMock({
                    state: {
                        configuration: { accessibility: false, can_create_program_increment: true },
                    },
                }),
            },
        };

        const wrapper = shallowMount(FeatureCard, component_options);
        expect(wrapper.element).toMatchSnapshot();
    });
});
