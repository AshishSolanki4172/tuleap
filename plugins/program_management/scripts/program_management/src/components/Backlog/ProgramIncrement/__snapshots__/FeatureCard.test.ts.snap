// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureCard Displays a draggable card with accessibility pattern 1`] = `
<div
  class="element-backlog-items"
  data-element-id="100"
  data-tlp-tooltip=""
  draggable="true"
>
  <div
    class="element-card element-card-lake_placid_blue element-draggable-item element-card-background-peggy_pink_text element-card-with-accessibility"
  >
    <div
      class="element-card-content"
    >
      <div
        class="element-card-xref-label"
      >
        <a
          class="element-card-xref element-card-xref-lake_placid_blue"
          data-not-drag-handle="true"
          href="/plugins/tracker/?aid=100"
        >
          
                    
                
        </a>
         
        <span
          class="element-card-label"
        >
          My artifact
        </span>
      </div>
    </div>
     
    <div
      class="element-card-accessibility"
    />
  </div>
   
  <!---->
</div>
`;

exports[`FeatureCard Displays a not draggable card when feature has planned user stories 1`] = `
<div
  class="element-backlog-items tlp-tooltip tlp-tooltip-left"
  data-element-id="100"
  data-tlp-tooltip="The feature has elements planned in team project, it can not be unplanned"
  draggable="false"
>
  <div
    class="element-card element-card-lake_placid_blue"
  >
    <div
      class="element-card-content"
    >
      <div
        class="element-card-xref-label"
      >
        <a
          class="element-card-xref element-card-xref-lake_placid_blue"
          data-not-drag-handle="true"
          href="/plugins/tracker/?aid=100"
        >
          
                    
                
        </a>
         
        <span
          class="element-card-label"
        >
          My artifact
        </span>
      </div>
    </div>
     
    <!---->
  </div>
   
  <!---->
</div>
`;

exports[`FeatureCard Displays a not draggable card when user can not plan/unplan features 1`] = `
<div
  class="element-backlog-items tlp-tooltip tlp-tooltip-left"
  data-element-id="100"
  data-tlp-tooltip="The feature is not plannable, user does not have permission to update artifact or field link."
  draggable="false"
>
  <div
    class="element-card element-card-lake_placid_blue"
  >
    <div
      class="element-card-content"
    >
      <div
        class="element-card-xref-label"
      >
        <a
          class="element-card-xref element-card-xref-lake_placid_blue"
          data-not-drag-handle="true"
          href="/plugins/tracker/?aid=100"
        >
          
                    
                
        </a>
         
        <span
          class="element-card-label"
        >
          My artifact
        </span>
      </div>
    </div>
     
    <!---->
  </div>
   
  <!---->
</div>
`;

exports[`FeatureCard Displays a not draggable card without accessibility pattern 1`] = `
<div
  class="element-backlog-items"
  data-element-id="100"
  data-tlp-tooltip=""
  draggable="true"
>
  <div
    class="element-card element-card-lake_placid_blue"
  >
    <div
      class="element-card-content"
    >
      <div
        class="element-card-xref-label"
      >
        <a
          class="element-card-xref element-card-xref-lake_placid_blue"
          data-not-drag-handle="true"
          href="/plugins/tracker/?aid=100"
        >
          
                    
                
        </a>
         
        <span
          class="element-card-label"
        >
          My artifact
        </span>
      </div>
    </div>
     
    <!---->
  </div>
   
  <!---->
</div>
`;
